### PR TITLE
[EPIC] Relaying

### PR DESCRIPTION
**Pull request by iamacook**
## What it solves

- [x] Adds relaying as an execution method. #1700 
- [x] Dashboard widget with relaying feature information. #1701 
   ~Relaying educational series~. #1704 
- [x] Create Safe #1802 
- [x] Batch transactions modal #1818 

## TODO

- [x] Separate `NEXT_PUBLIC_SAFE_GELATO_RELAY_SERVICE_URL` into production and staging.

## How this PR fixes it
- Enables the transaction execution via a sponsored relayer.
- Displays the number of remaining hourly relays.
- Smart Contract signers can only relay a fully signed transaction.
- In a relaying execution, only the `Gas limit` can be modified.
- New widget in the dashboard containing relaying information and a link to the educational series
~- New route with the relaying educational series~
- Safe creation will be relayed if all Safe signers have relay transactions available
- Relaying is also enabled in the Batch transactions modal

## How to test it
_Execution flow_
1. In Goerli, get to the review screen of a transaction that can be executed by the connected account.
2. If the "Execute transaction" checkbox is ticked, you are able to see a menu to chose the _execution method_
3. On submitting the transaction "Via relayer" the transaction will move through the stages RELAYING -> INDEXING -> SUCCESS in the happy case.

~_Educational series_~
~1. From the dashboard, click in "Read about trial" in the new **Gasless transactions** widget.~
~2. You'll be able to navigate the relaying educational series.~

_Batch transactions_
1. Fully sign 2+ transactions so that you can batch them
2. The "How to pay" info should be present in the Modal
3. Click "Send" will relay the Multisend transaction

_Safe creation_
1. Enter the create Safe flow
2. In the review step, if all the Safe owners have available relaying transaction, a component with that info is displayed
3. On clicking "Next", the Safe creation will be relayed.
4. The transaction will be monitored until being indexed.

## Screenshots
Screenshot 2023-03-31 at 11 28 39
Screenshot 2023-03-31 at 11 29 06
Screenshot 2023-03-31 at 11 29 19

## Checklist
* [x] I've tested the branch on mobile 📱
* [ ] I've documented how it affects the analytics (if at all) 📊
* [ ] I've written a unit/e2e test for it (if applicable) 🧑‍💻

### CODE REVIEW

**Comment by liliya-soroka** [comment]
@DiogoSoaress , could you check the limitation per 5 tx? For me, it jumped from 4 to 1 without 3,2...And now I have 0 available tx, but on execution modal opening either 1 available or 0 available is displayed.
https://drive.google.com/file/d/1R2DVV3msp_Ad5xPTy1IjIr9kTmOxHYIi/view?usp=sharing

Could the number of available txs depend on the executor?

**Comment by DiogoSoaress** [comment]
Can you check the network request @liliya-soroka ? The UI should be very dumb.

> Could the number of available txs depend on the executor?
No, it should be per safe address.

**Comment by francovenica** [comment]
The values in adv parameters should not be editable when the relaying option is checked (and they are not), but you can edit them while selecting the connected wallet and then change to the relayer with the new edited values, defeating the purpose of graying out the inputs in the first place

I'd suggest that as soon as you switch to the relayer the values in adv parameters are set to the ones estimated in the beginning

relayer

**Comment by francovenica** [comment]
I tried some tx and the notification ends up saying that the tx failed, but the tx itself was successfully executed
image
image
https://epic_relaying_v1--webcore.review-web-core.5afe.dev/transactions/tx?id=multisig_0x0EDc6697427E9e9d62A1eE1A2951AfA7bE9e75B7_0x04a36e94f4f01aac6be324344b277475dd45b09905aa96575709a38c5eccd5d8&safe=gor%3A0x0EDc6697427E9e9d62A1eE1A2951AfA7bE9e75B7

**Comment by francovenica** [comment]
Looks good.
I tried to signed with MM in chrome, MM in phone app (via WC), Safe mobile pairing, trezor one, ledger nano S and safe-as-an-owner.
I used safes 1/x and 2/x, trying send funds, owner edition, apps (tx builder with dummy contracts) and setting spending limit (setting them, using them do not use the relayer).
I made sure that contracts cannot use the relayer unless the tx is fully signed first (with safes-as-an-owner)

So far so good, I'll take another look when the safe creation and batch are deployed in the epic

**Comment by francovenica** [comment]
Question regarding the frequency in which you can create safes and be paid by the relayer
I assumed that was based on the address connected, but it seems it depends on the type of wallet connected. 
In the following gif first I'm connected with the MM phone app through WC, which i used to create a bunch of safes, but then I change to the MM chrome extension, that uses the same address, and now I have my safe paid by the relayer:
relayer counter

**Comment by francovenica** [comment]
Note: there is an old issue that I'm pretty sure has nothing to do with this PR. 
When creating safes with the stg CGW, after submition the safe steps would get stuck in "Processing" and it will tell you that you can use the safe, but whenever you enter you will see that the safe cannnot be loaded. 
image


Because of this I'm currently using Prod CGW

**Comment by francovenica** [comment]
Issue:
I can only execute the tx batch once. For some reason I can execute 2 tx in batch, but after that, any attempt to execute another batch is always rejected by the relayer

In this safe I was able to add 2 new owners, then I tried to remove them both in a batch tx, but is always rejected by the relayer https://epic_relaying_v1--webcore.review-web-core.5afe.dev/transactions/queue?safe=gor:0x8E1c75aBa457c97EfA9BC97BA8C50A09361AA202

image

**Comment by DiogoSoaress** [comment]
> Question regarding the frequency in which you can create safes and be paid by the relayer I assumed that was based on the address connected, but it seems it depends on the type of wallet connected. In the following gif first I'm connected with the MM phone app through WC, but then I change to the MM chrome extension, that uses the same address, and now I have my safe paid by the relayer: relayer counter [ relayer counter ](https://user-images.githubusercontent.com/4503659/229009186-cc6dd2d1-c86d-4c8a-a3a1-a09a5cb1be1d.gif) [ ](https://user-images.githubusercontent.com/4503659/229009186-cc6dd2d1-c86d-4c8a-a3a1-a09a5cb1be1d.gif)

The rate limit on creation is per Safe owner: you will have the least remaining relays across the chosen owners.

Example - if you select an owner that has 0 relays available, you won't be able to relay the Safe creation.

**Comment by DiogoSoaress** [comment]
> Note: there is an old issue that I'm pretty sure has nothing to do with this PR. When creating safes with the stg CGW, after submition the safe steps would get stuck in "Processing" and it will tell you that you can use the safe, but whenever you enter you will see that the safe cannnot be loaded. image
> 
> Because of this I'm currently using Prod CGW

You're right, this happens while the Transaction Service hasn't indexed the created Safe. The Safe is created but can't be fetched yet.

File: .env.example [hunk 1]
```diff
@@ -25,4 +25,7 @@ NEXT_PUBLIC_BEAMER_ID=
 NEXT_PUBLIC_WC_BRIDGE=
 
 # E2E tests
-NEXT_PUBLIC_CYPRESS_MNEMONIC=
\ No newline at end of file
+NEXT_PUBLIC_CYPRESS_MNEMONIC=
+
+# Safe Gelato relay service
+NEXT_PUBLIC_SAFE_GELATO_RELAY_SERVICE_URL=
```
Comment (katspaugh) [inline .env.example:31]: We should add this to the readme as well.

File: public/images/common/relayer.svg [hunk 2]
```diff
@@ -0,0 +1,24 @@
+<svg width="24" height="25" viewBox="0 0 24 25" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (katspaugh) [inline public/images/common/relayer.svg:1]: Is this image still used?

Comment (DiogoSoaress) [reply public/images/common/relayer.svg:1]: Good catch, I forgot to remove it

**Comment by liliya-soroka** [comment]
please , make sure that the prod rtelay-service is used with prod CGW . Seems that we are using staging relay -service env with prod CGW now

**Comment by francovenica** [comment]
> > Question regarding the frequency in which you can create safes and be paid by the relayer I assumed that was based on the address connected, but it seems it depends on the type of wallet connected. In the following gif first I'm connected with the MM phone app through WC, but then I change to the MM chrome extension, that uses the same address, and now I have my safe paid by the relayer:
> 
> The rate limit on creation is per Safe owner: you will have the least remaining relays across the chosen owners.
> 
> Example - if you select an owner that has 0 relays available, you won't be able to relay the Safe creation.

I don't understand this. While you are creating a safe you are not the owner of anything. I'd assume that is the address you are currently using that is limited to 5 safes per hour. Is that how it should work?
If that is the case, then here we have a problem. In the gif I showed you the same owner could create safes while being connected with the MM chrome extension but could not when connected with the MM phone app via WC. 
I'd assume that if I use the same address connected by any means (MM chrome extension, MM phone app, Safe Mobile pairing...) it would have the same amount of safe to create no matter how he is connected to the app

**Comment by francovenica** [comment]
The issue i had in [this](https://github.com/safe-global/web-core/pull/1776#issuecomment-1491247490) comment keeps comming back. This time I waited for several hours before tying to execute the tx again. Tried 4 times already and I always have the same issue

I add 2 owners by batch tx and then I try to remove them both at the same time again through a batch tx. 
I've been able to remove owners through batch tx before, so I don't think is the type of tx (owner edition)

The safe (the tx are still there, try to execute them)
https://epic_relaying_v1--webcore.review-web-core.5afe.dev/transactions/queue?safe=gno:0x3F8c62c9581c8cbA6E4AFf36931c597aB95a85Ff

Snapshots:
image

The full error in the snapshot:
```
Error: Relayed transaction was cancelled by relay provider.
    at https://epic_relaying_v1--webcore.review-web-core.5afe.dev/_next/static/chunks/pages/_app-226919d57ade70c4.js:274:1438224
    at Array.forEach (<anonymous>)
    at https://epic_relaying_v1--webcore.review-web-core.5afe.dev/_next/static/chunks/pages/_app-226919d57ade70c4.js:274:1438162
    at s (https://epic_relaying_v1--webcore.review-web-core.5afe.dev/_next/static/chunks/main-0bdfd69a1d734bc0.js:1:114987)
    at Generator._invoke (https://epic_relaying_v1--webcore.review-web-core.5afe.dev/_next/static/chunks/main-0bdfd69a1d734bc0.js:1:114775)
    at P.forEach.e.<computed> [as next] (https://epic_relaying_v1--webcore.review-web-core.5afe.dev/_next/static/chunks/main-0bdfd69a1d734bc0.js:1:115410)
    at r (https://epic_relaying_v1--webcore.review-web-core.5afe.dev/_next/static/chunks/pages/_app-226919d57ade70c4.js:274:1724943)
    at s (https://epic_relaying_v1--webcore.review-web-core.5afe.dev/_next/static/chunks/pages/_app-226919d57ade70c4.js:274:1725146)
```

**Comment by schmanu** [comment]
@francovenica 

> he issue i had in https://github.com/safe-global/web-core/pull/1776#issuecomment-1491247490 comment keeps comming back. This time I waited for several hours before tying to execute the tx again. Tried 4 times already and I always have the same issue
> I add 2 owners by batch tx and then I try to remove them both at the same time again through a batch tx.
I've been able to remove owners through batch tx before, so I don't think is the type of tx (owner edition)

This is to be expected. The owners are stored on-chain in a forward-linked-list. To remove an owner you pass in the `prevOwner` in that list such that we can efficiently link the prevOwner to the nextOwner which effectively removes the owner from that list.

In your transactions, the first tx will remove `gno:0x6E45d69a383CECa3d54688e833Bd0e1388747e6B` as owner.
The second tx will then remove `gno:0x7724b234c9099C205F03b458944942bcEBA13408` and that transaction passes 0x6E45d69a383CECa3d54688e833Bd0e1388747e6B as `prevOwner`.
Screenshot 2023-04-03 at 10 21 25

These transactions will always fail as a batch. It is unrelated to relaying.

For testing batched tx happy paths I would suggest leaving policy / settings changes out of batches as they are likely to cause some trouble.

**Comment by francovenica** [comment]
I tried in this pr, in the ETH network so no relayer. I was able to execute a batch tx removing 2 owners with no issues. Maybe is important in which order you remove them?
image

**Comment by schmanu** [comment]
> I tried in this pr, in the ETH network so no relayer. I was able to execute a batch tx removing 2 owners with no issues. Maybe is important in which order you remove them?

I depends on the order of the addresses you are going to remove. In the screenshot you attached in the working example, none of the owners that are getting removed appear as `prevOwner` of another transaction.

It will always depend on the order of owners. In the (relayed) example you posted on Görli it will not be possible. In the example on mainnet it will always be possible. But in either case it is not related to relaying.

**Comment by francovenica** [comment]
TL:DR; the batch tx issue is not about the relayer. The relayer works fine, we can create the RC

Ok, given that the issue with the remove owner with batch tx is not a relayer problem we can ignore it for this ticket at least (Aaron told me that fixing that issue was already disccussed to be done later)

The safe creation works fine. 
The counter of tx sponsored depends on the owners the safe are including during its creation. 
If several owner are set, the counter will take in account the owner with the lowest counter so far
Editing the owner once the safe created doesn't affect the counter during safe creation
Creating a safe normally (with no relayer) works fine as well

Executing txs in batch works fine
Tried several types of tx and they work fine
Tx without the relayer works fine as well

**Comment by iamacook** [comment]
I've just verified the prod. API keys working as well.

File: src/components/tx/SponsoredBy/index.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    <Box className={css.sponsoredBy}>
+      <SvgIcon component={GasStationIcon} inheritViewBox className={css.icon} />
+      <Stack direction="column">
+        <Stack direction="row" spacing={0.5} alignItems="center" mb={1}>
+          <Typography variant="body2" fontWeight={700} letterSpacing="0.1px">
+            Sponsored by
+          </Typography>
+          <img src="/images/common/gnosis-chain-logo.png" alt="Gnosis Chain" className={css.gcLogo} />
+          <Typography variant="body2" fontWeight={700} letterSpacing="0.1px">
+            Gnosis Chain
```
Comment (katspaugh) [inline src/components/tx/SponsoredBy/index.tsx:19]: Please don't hardcode stuff like this in future.

File: src/hooks/useRemainingRelays.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    const res = await fetch(url)
+    const data = await res.json()
+    return data.remaining
+  } catch (error) {
+    logError(Errors._630, (error as Error).message)
+    return 0
+  }
+}
+
+export const useRemainingRelaysBySafe = () => {
```
Comment (katspaugh) [inline src/hooks/useRemainingRelays.ts:31]: Please always add return types to functions. Although it will likely infer it, it's easier for humans to read when a function's signature is typed.

File: src/hooks/useWalletCanRelay.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    if (!tx || !wallet) return
+
+    return isSmartContractWallet(wallet)
+      .then((isSCWallet) => {
+        if (!isSCWallet) return true
+
+        return hasEnoughSignatures(tx, safe)
+      })
+      .catch((err) => {
+        console.error("Couldn't determine if a Smart Contract wallet is connected", err)
```
Comment (katspaugh) [inline src/hooks/useWalletCanRelay.ts:22]: Use `logError`, not console.error directly.

File: src/services/tx/sponsoredCall.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    if (res.ok) {
+      return res.json()
+    }
+    return res.json().then((data) => {
+      throw new Error(`${res.status} - ${res.statusText}: ${data?.error?.message}`)
+    })
+  })
+
+  return data
+}
```
Comment (katspaugh) [inline src/services/tx/sponsoredCall.ts:31]: So there's a service for relayed executions, but fetching remaining relays is just a function in some other hook...

File: src/components/tx/SignOrExecuteForm/index.tsx [hunk 7]
```diff
@@ -156,19 +174,47 @@ const SignOrExecuteForm = ({
...
+
+      id = await proposeTx(signedTransaction)
+      safeTx = signedTransaction
+    }
+
+    if (!id) {
+      throw new Error('Transaction could not be proposed')
+    }
+
+    trackEvent({ ...MODALS_EVENTS.PROPOSE_TX, label: ExecutionType.RELAYER })
```
Comment (katspaugh) [inline src/components/tx/SignOrExecuteForm/index.tsx:206]: Tracking right in the modal is no bueno. We have a tx tracking hook based on tx events.

File: src/services/tx/sponsoredCall.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import { SAFE_GELATO_RELAY_SERVICE_URL } from '@/hooks/useRemainingRelays'
+import { type SafeTransactionData } from '@safe-global/safe-core-sdk-types'
+
+// TODO: import type from relay-service
```
Comment (katspaugh) [inline src/services/tx/sponsoredCall.ts:4]: Why was this never done?

File: src/hooks/useRemainingRelays.ts (hunk 4 above)
Comment (iamacook) [reply src/hooks/useRemainingRelays.ts:31]: Return types speed up type checking as well.

**Review by katspaugh — APPROVED** [review]
Approving since this was already tested, but the PR comments should be addressed in #1838.
